move pivot_table doc-string to DataFrame

jbrockmendel opened the pull request:
instead of pinning it on over pin `core.reshape.pivot`

At the moment `core.reshape.pivot.pivot_table` is pinned to `DataFrame` by writing `DataFrame.pivot_table = pivot_table`.  This PR defines `pivot_table` directly in `DataFrame` and then calls to the function.  It's the same pattern already used by `DataFrame.pivot`.

I think this is the last `DataFrame` method pinned on to `DataFrame` this way.  Two more left in `Series`

 - [ ] closes #xxxx
 - [ ] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [ ] whatsnew entry

jbrockmendel committed:
Define pivot_table in DataFrame instead of

pinning it on over pin reshape.pivot

pep8speaks commented:
Hello @jbrockmendel! Thanks for submitting the PR.



 - In the file [`pandas/core/reshape/pivot.py`](https://github.com/pandas-dev/pandas/blob/773c5b7705eaf11ba1ca3adff3c42a01e68acfda/pandas/core/reshape/pivot.py), following are the PEP8 issues :

> [Line 130:1](https://github.com/pandas-dev/pandas/blob/773c5b7705eaf11ba1ca3adff3c42a01e68acfda/pandas/core/reshape/pivot.py#L130): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 0

jreback reviewed (changes requested):
if u look iin the git history we just moved this
to be this way

why are you trying to move back?

jbrockmendel commented:
> why are you trying to move back?

Because pinning methods/attributes on from afar makes these classes harder to reason about.  In a non-trivial set of cases, these classes are altered at import time with `hasattr` checks, making the code brittle to changes in import order (#16931, #16959).  There is One Obvious Way To Do It, and that's to define the method in the class definition.

I'll take a look at the old discussion and try to understand the reasoning.

jorisvandenbossche commented:
+1 on this change. I would only keep the docstring with the implementation, and add it on the dataframe one.

jreback commented:
@jorisvandenbossche you are not remembering: https://github.com/pandas-dev/pandas/pull/15521/files

jbrockmendel commented:
> I would only keep the docstring with the implementation, and add it on the dataframe one.

That was my initial inclination too, but then attaching the docstring to `DataFrame.pivot_table` causes a circular import mess.  `core.reshape.pivot` imports `DataFrame` and a bunch of other modules, some of which also import `DataFrame`.  I'd like to de-tangle this, but trying to do that in this PR would snowball.

jreback commented:
@jbrockmendel ``pandas/core/frame.py`` is WAY too long. Functions are often defined externally to better group them on purpose.

Moving things back defeats the purpose of delegation and simplicity. The maxims of python need to be balanced with practicality.

This is an anti-change to all pandas philosophy.

jbrockmendel commented:
> `pandas/core/frame.py` is WAY too long

Very much agreed.  

How about this: I'll implement @jorisvandenbossche 's suggestion of putting the docstring with the implementation in `core.reshape.pivot` (or in `_shared_docs`).  Then defining the method directly in `frames.DataFrame` will take up 9 lines (4 if you don't count wrapping).  While I'm at it, I'll do the same for `DataFrame.pivot.__doc__`, `_shared_docs['melt']`, and `DataFrame._agg_doc` (which I expect doesn't belong in the namespace anyway).

Then `core.frames` is a few hundred lines shorter, tangled imports in `core.reshape` are replaced with `ABCFoo` imports, and the original goal of the PR is accomplished.

jorisvandenbossche commented:
> @jorisvandenbossche you are not remembering: https://github.com/pandas-dev/pandas/pull/15521/files

Not sure what you are trying to argue .. as the PR did exactly the same as this PR is now doing for another method?

jorisvandenbossche commented:
> Moving things back

@jreback this PR is *not* moving the actual *implementation* back, it is just changing the way how the the method is added to the DataFrame class to a more explicit and consistent (with other methods) way.

jbrockmendel committed:
Fix NameError caused by referencing DataFrame incorrectly;

avoid import of DataFrame altogether using ABC classes where possible

jbrockmendel committed:
Fix import from wrong module

jbrockmendel committed:
Fix broken test by checking isinstance(by, Index) instead of ABCIndex

ABCIndex appears to not recognize Int64Index, needs an Issue

jbrockmendel commented:
The test failures are being caused by a problem with `ABCIndex`.  I've fixed them for this PR and am about to open an issue for the underlying problem.

jreback commented on pandas/core/reshape/pivot.py:
```diff
@@ -16,81 +18,7 @@
-    DataFrame.pivot : pivot without aggregation that can handle
-        non-numeric data
-    """
+    """ See DataFrame.pivot_table.__doc__ """
```
in the future, pls don't put more than 1 change in a PR. These import changes should be separate and are orthogonal to this change.

jbrockmendel replied:
OK.

jreback reviewed (changes requested):
this doc-string needs to go in _shared_docs; then you need to add it back to pivot_table itself, which is imported as ``pd.pivot_table``.

jbrockmendel committed:
Move pivot_table.__doc__ into _shared_docs

jbrockmendel commented:
> this doc-string needs to go in _shared_docs; then you need to add it back to pivot_table itself, which is imported as pd.pivot_table.

Just pushed a commit to to do this.

You mentioned frame.py being too long, which is totally correct.  Would it be helpful to do the same thing with the docstrings for `pivot`, `melt`, `unstack`?

jbrockmendel committed:
Add pivot_table to _shared_docs in frame,

then import frame before reshape.pivot.  Going in the opposite order
results in a circular import

jbrockmendel committed:
Fix missing import if pivot_table

jreback commented:
> You mentioned frame.py being too long, which is totally correct. Would it be helpful to do the same thing with the docstrings for pivot, melt, unstack?

jorisvandenbossche commented:
@jreback You closed by accident?

jreback commented:
yes sorry

jreback commented on pandas/core/reshape/pivot.py:
```diff
@@ -2,95 +2,28 @@
 from pandas.core.reshape.concat import concat
-from pandas import Series, DataFrame, MultiIndex, Index
+from pandas.core.frame import _shared_docs
+# Note: We need to make sure `frame` is imported before `pivot`, otherwise
```
hmm I don't like this here the import order shouldn't matter (I know it does, but it shouldn't).

generally we simply import the _shared_docs at the end.

jreback reviewed (changes requested):
see comment

jbrockmendel commented:
> hmm I don't like this here the import order shouldn't matter (I know it does, but it shouldn't).

Yah, I very much dislike this part.  My preference would be to put the docstring in `reshape.pivot` and refactor imports in `reshape.pivot` so that `core.frame` can import it at the top.  This spills over into requiring the same import changes in `reshape.concat`, so I figured it would be better to get the go-ahead before trying this (in a separate PR).

> generally we simply import the _shared_docs at the end.

Coming right up

jbrockmendel committed:
Move import of _shared_docs to end of reshape.pivot imports

jbrockmendel commented:
btw, why isn't `_shared_docs` shared between `core.base` and `core.generic`?

jreback commented:
> btw, why isn't _shared_docs shared between core.base and core.generic?

its not important for base, rather its sub-class based
